cypress-io/cypress
# fix: save spec filter term

- Closes https://github.com/cypress-io/cypress/issues/21837

### User facing changelog

Cypress will remember the spec filter in the "Search Specs" field during your session, and also when a project is reopened. The same filter will be used in the main specs list as well as the inline specs list present in the Test Runner.

### Additional details

Previously, the search term used to find specs was not preserved when you navigated away from the specs list, and was not carried over between the inline specs list and the main specs list. This was a source of frustration for users, especially who are used to this behavior from before Cypress 10.

In earlier versions of Cypress, this term was saved in local storage in the launchpad Electron app. That approach doesn't work in Cypress 10 because the filtering now happens in the browser app, so users would be saving different search terms depending on which browser they are using at the time. So this PR adds the filter term to the projects `savedState` where some other project-specific information is already stored.

Some notes on the implementation:

1. A `specFilter` field has been added to `specs-store.ts` to preserve it when the user has left the Specs List page. This is the source of truth for the value of the filter on the client side.
2. A `useSpecFilter` composable is used in `SpecsList.vue` and `InlineSpecList.vue` which handles initializing the value in the store (from the saved data in the `currentProject.savedState.specFilter` field in gql), updating the store as the user types, and saving the search term back to gql. This reduces some of the duplication between these files and means we have a single place to, eg, set the the debounce time.
3. The existing `setPreferences` functionality has been extended to allow `global` and `project` preferences to be set. "Global" preferences are saved in the `__project__/state.json` file, and "project" preferences are saved in the `state.json` found the in the unique directory for that project. This required updating the existing usages of `setPreferences` and some related tests that stubbed it.
4. Created a new issue to clean up the confusing use of the `cache` file for project data like browser and testing type, those things probably doesn't need to live there. #23202

I looked at avoiding the store and driving this fully through graphql, but it seemed to require about the same amount of work, and using the store matches a little better with how we handle global preferences already.

We could optimize further by refactoring `setPromptShown` as well but we're about to be right back in this code, let's keep this PR from getting any bigger and make other changes in new PRs.

### Steps to test

- Open any project and search for a spec
- Refresh the page and see the search term is still there
- Navigate to other pages on the app and go back to the specs list, see the search term is still there
- Open a spec and open the inline specs list, see the search term is prefilled there
- Refresh the spec page to make sure the search term prefills based on the saved term in gql
- Type a new search in the Inline Specs List and go back to the main specs list, the new search should be there as well
- Open a different project, it should not have a saved search term.
- Search for a spec in the newly opened project, then close it and go back to to the first project, and see that the search terms saved are specific to the projects.
- Use the X in the specs list, and the clear button in the No Results state to confirm clearing the search in these ways also clears the saved search.

### How has the user experience changed?

The Spec filter term was previously discarded when users navigated away from the page they were using to search. Now the filter is saved with the project and is only removed by user action.

### Percy

Note: changing some test nesting has changed the names of percy snapshots, so some will appear to be net-new that really aren't.

### PR Tasks


- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? This isn't currently documented, so no change needed.
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
